pkgs/by-name/py/pyfa/package.nix

```
@@ -1,31 +1,124 @@
 {
   lib,
-  appimageTools,
-  fetchurl,
+  python3Packages,
+  fetchFromGitHub,
+  gsettings-desktop-schemas,
+  adwaita-icon-theme,
+  wrapGAppsHook3,
+  gdk-pixbuf,
 }:
 let
-  version = "2.60.2";
+  version = "2.61.0";
 in
-appimageTools.wrapType2 {
+python3Packages.buildPythonApplication rec {
   inherit version;
   pname = "pyfa";
+  format = "other";
 
-  src = fetchurl {
-    name = "pyfa-appimage-${version}";
-    url = "https://github.com/pyfa-org/Pyfa/releases/download/v${version}/pyfa-v${version}-linux.AppImage";
-    hash = "sha256-6doetQ6E1Occ/SqewfxRqPEX1MnUuFomm+8VmetIz4Y=";
+  src = fetchFromGitHub {
+    owner = "pyfa-org";
+    repo = "Pyfa";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-VSuRQugUODc+LbhKbzsA09pnqPEIAt0pQS2An/p7r9A=";
   };
 
+  build-system = [ python3Packages.setuptools ];
+  dependencies = with python3Packages; [
+    wxpython
+    logbook
+    matplotlib
+    python-dateutil
+    requests
+    sqlalchemy_1_4
+    cryptography
+    markdown2
+    beautifulsoup4
+    pyaml
+    roman
+    numpy
+    python-jose
+    requests-cache
+  ];
+
+  buildInputs = [
+    gsettings-desktop-schemas
+    adwaita-icon-theme
+    gdk-pixbuf
+  ];
+
+  dontWrapGApps = true;
+  nativeBuildInputs = [
+    python3Packages.pyinstaller
+    wrapGAppsHook3
+  ];
+
+  #
+  # upstream does not include setup.py
+  #
+  patchPhase = ''
+    cat > setup.py <<EOF
+      from setuptools import setup
+      setup(
+        name = "${pname}",
+        version = "${version}",
```

instead of escaping `EOF` you could
```suggestion
        name = "$pname",
        version = "$version",
```
in turn making the version overrideable

---

searching nixos/nixpkgs for $pname declaration without braces, I don't see any adoption of this practice;

maybe a good write up on it would help understanding the benefit and increasing the adoption of the pratice?

---

In Nix derivations, attributes are automatically passed to the builder script as Bash variables. You can try running echo $pname yourself and see that it exists and it is set to the current value of pname.

The main benefit of this theoretically is that when a package is overridden with overrideAttrs, the setup.py reacts to the override as it is synchronized with the overridden variables. Using string substitution will bake the values directly into the builder, meaning that an override won't change the generated setup.py.

*That said,* I'm not 100% certain that this will cause a rebuild when pname and version are overridden, unlike using other override-aware mechanisms like `finalAttrs`. Try if it works first

---

https://matt-rickard.com/avoiding-footguns

If you want a silly example:

> [!CAUTION]
> DO NOT RUN THIS

```shell
cat <<EOF >foobar.py
# prevent use of `sudo rm -rf / --no-preserve-root`
with open(file, "r") as f:
    data = f.read()
data = data.replace(
    "sudo rm -rf / --no-preserve-root",
    "echo sudo rm -rf / --no-preserve-root",
)
with open(file, "w") as f:
    f.write(data)
EOF
```


Avoiding footguns by default should never be considered bad practice. The code might be fine now, but a later contributor may trigger an unwanted case

> searching nixos/nixpkgs for $pname declaration without braces, I don't see any adoption of this practice

true, we actively try and avoid using `pname`.

```suggestion
        name = "pyfa",
        version = "$version",
```

but personally still prefer using `'EOF'`

> Try if it works first

`buildPythonApplication` does not yet support the `finalAttrs` pattern AFAIK.